Mathlib/Algebra/Order/Archimedean/Class.lean

```diff
@@ -452,3 +452,127 @@ theorem liftOrderHom_mk (f : M → α) (h : ∀ a b, mk a ≤ mk b → f a ≤ f
+
+@[to_additive]
+instance [MulArchimedean M] : Subsingleton (MulArchimedeanClass₁ M) where
+  allEq A B := by
+    induction A using ind with | mk a ha
+    induction B using ind with | mk b hb
+    simpa [mk] using MulArchimedeanClass.mk_eq_mk_of_mulArchimedean ha hb
+
+@[to_additive]
+instance [Nontrivial M] : Nonempty (MulArchimedeanClass₁ M) := by
+  obtain ⟨x, hx⟩ := exists_ne (1 : M)
+  exact ⟨mk hx, by simpa using hx⟩
+
+/-- Lift a `f : {a : M // a ≠ 1} → α` function to `MulArchimedeanClass₁ M → α`. -/
+@[to_additive "Lift a `f : {a : M // a ≠ 0} → α` function to `ArchimedeanClass₀ M → α`."]
+def lift {α : Type*} (f : {a : M // a ≠ 1} → α)
+    (h : ∀ (a b : {a : M // a ≠ 1}), mk a.prop = mk b.prop → f a = f b) :
+    MulArchimedeanClass₁ M → α := fun ⟨A, hA⟩ ↦ by
+  refine (MulArchimedeanClass.lift
+    (fun b ↦ if h : b = 1 then ⊤ else WithTop.some (f ⟨b, h⟩)) ?_ A).untop ?_
+  · intro a b h'
+    simp only
+    split_ifs with ha hb hb
+    · rfl
+    · rw [ha] at h'
+      exact (hb (MulArchimedeanClass.mk_eq_top_iff.mp h'.symm)).elim
+    · rw [hb] at h'
+      exact (ha (MulArchimedeanClass.mk_eq_top_iff.mp h')).elim
+    · exact WithTop.coe_eq_coe.mpr <| h ⟨a, ha⟩ ⟨b, hb⟩ (by simpa [mk] using h')
+  · induction A using MulArchimedeanClass.ind with | mk a
+    simpa using MulArchimedeanClass.mk_eq_top_iff.not.mp hA
+
+@[to_additive (attr := simp)]
+theorem lift_mk {α : Type*} (f : {a : M // a ≠ 1} → α)
+    (h : ∀ (a b : {a : M // a ≠ 1}), mk a.prop = mk b.prop → f a = f b) {a : M} (ha : a ≠ 1) :
+    lift f h (mk ha) = f ⟨a, ha⟩ := by simp [lift, mk, ha]
+
+/-- Lift a function `{a : M // a ≠ 1} → α` that's monotone along archimedean classes to a
+monotone function `MulArchimedeanClass₁ M →o α`. -/
+@[to_additive "Lift a function `{a : M // a ≠ 1} → α` that's monotone along archimedean classes to a
+monotone function `ArchimedeanClass₀ M₁ →o α`."]
+noncomputable
+def liftOrderHom {α : Type*} [PartialOrder α]
+    (f : {a : M // a ≠ 1} → α) (h : ∀ (a b : {a : M // a ≠ 1}), mk a.prop ≤ mk b.prop → f a ≤ f b) :
+    MulArchimedeanClass₁ M →o α where
+  toFun := lift f fun a b heq ↦ le_antisymm (h a b heq.le) (h b a heq.ge)
+  monotone' A B hAB := by
+    induction A using ind with | mk a ha
+    induction B using ind with | mk b hb
+    simpa using h ⟨a, ha⟩ ⟨b, hb⟩ hAB
+
+@[to_additive (attr := simp)]
+theorem liftOrderHom_mk {α : Type*} [PartialOrder α]
+    (f : {a : M // a ≠ 1} → α) (h : ∀ (a b : {a : M // a ≠ 1}), mk a.prop ≤ mk b.prop → f a ≤ f b)
+    {a : M} (ha : a ≠ 1) : liftOrderHom f h (mk ha) = f ⟨a, ha⟩ :=
+  lift_mk f (fun a b heq ↦ le_antisymm (h a b heq.le) (h b a heq.ge)) ha
+
+variable (M) in
+/-- Adding top to the type of non-top classes yields the type of all classes. -/
+@[to_additive "Adding top to the type of non-top classes yields the type of all classes."]
+noncomputable
+def withTopOrderIso : WithTop (MulArchimedeanClass₁ M) ≃o MulArchimedeanClass M where
```

Comment:
Can you deduce this from `WithTop {x : α // x ≠ ⊤} ≃o α`? And please also add a `WithBot` version